cores/esp8266/core_esp8266_waveform.c

```diff
@@ -38,84 +38,41 @@
 */
 
 #include <Arduino.h>
+#include "ets_sys.h"
 #include "core_esp8266_waveform.h"
 
-// Need speed, not size, here
-#pragma GCC optimize ("O2")
-
 // Maximum delay between IRQs
 #define MAXIRQUS (10000)
 
-// If the cycles from now to an event are below this value, perform it anyway since IRQs take longer than this
-#define CYCLES_FLUFF (100)
-
-// Macro to get count of predefined array elements
-#define countof(a) ((size_t)(sizeof(a)/sizeof(a[0])))
-
-// Set/clear *any* GPIO
-#define SetGPIOPin(a) do { if (a < 16) { GPOS |= (1<<a); } else { GP16O |= 1; } } while (0)
-#define ClearGPIOPin(a) do { if (a < 16) { GPOC |= (1<<a); } else { GP16O &= ~1; } } while (0)
-// Set/clear GPIO 0-15
+// Set/clear GPIO 0-15 by bitmask
 #define SetGPIO(a) do { GPOS = a; } while (0)
 #define ClearGPIO(a) do { GPOC = a; } while (0)
 
+// Convert from us to ESP8266 clocks
+#define MicrosecondsToCycles(m) (clockCyclesPerMicrosecond() * (m))
+
 // Waveform generator can create tones, PWM, and servos
 typedef struct {
-  uint32_t nextServiceCycle;        // ESP cycle timer when a transition required
-  uint32_t timeLeftCycles;          // For time-limited waveform, how many ESP cycles left
-  const uint16_t gpioMask;          // Mask instead of value to speed IRQ loop
-  const uint16_t gpio16Mask;        // Mask instead of value to speed IRQ loop
-  unsigned state              : 1;  // Current state of this pin
-  unsigned nextTimeHighCycles : 31; // Copy over low->high to keep smooth waveform
-  unsigned enabled            : 1;  // Is this GPIO generating a waveform?
-  unsigned nextTimeLowCycles  : 31; // Copy over high->low to keep smooth waveform
+  uint32_t nextServiceCycle;   // ESP cycle timer when a transition required
+  uint32_t expiryCycle;        // For time-limited waveform, the cycle when this waveform must stop
+  uint32_t nextTimeHighCycles; // Copy over low->high to keep smooth waveform
+  uint32_t nextTimeLowCycles;  // Copy over high->low to keep smooth waveform
 } Waveform;
 
 // These can be accessed in interrupts, so ensure to bracket access with SEI/CLI
-static Waveform waveform[] = {
-  {0, 0, 1<<0, 0, 0, 0, 0, 0}, // GPIO0
-  {0, 0, 1<<1, 0, 0, 0, 0, 0}, // GPIO1
-  {0, 0, 1<<2, 0, 0, 0, 0, 0},
-  {0, 0, 1<<3, 0, 0, 0, 0, 0},
-  {0, 0, 1<<4, 0, 0, 0, 0, 0},
-  {0, 0, 1<<5, 0, 0, 0, 0, 0},
-  // GPIOS 6-8 not allowed, used for flash
-  // GPIO 9 and 10 only allowed in 2-bit flash mode
-#if !isFlashInterfacePin(9)
-  {0, 0, 1<<9, 0, 0, 0, 0, 0},
-  {0, 0, 1<<10, 0, 0, 0, 0, 0},
-#endif
-  // GPIO 11 not allowed, used for flash
-  {0, 0, 1<<12, 0, 0, 0, 0, 0},
-  {0, 0, 1<<13, 0, 0, 0, 0, 0},
-  {0, 0, 1<<14, 0, 0, 0, 0, 0},
-  {0, 0, 1<<15, 0, 0, 0, 0, 0},
-  {0, 0, 0, 1, 0, 0, 0, 0}  // GPIO16
-};
+static Waveform waveform[17];        // State of all possible pins
```

Comment:
The isFlashInterfacePin(9) macro no longer applies here?

Reply:
No, that's taken care of in `startWaveform` where we return `false` if it's an invalid pin or `isFlashInterfacePin()`.  The simpler array makes for simpler (faster) code in the IRQ.